[Eola-Z] iP

Eola-Z opened the pull request:
## AND HIS NAME IS....
# JOHNATHAN CENATOR
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/113225702/08fd22ab-fe17-4602-9bff-95a0f7325e69)
### YOU'LL NEVER SEE HIM COMING!!! 
You want a chatbot that can help you organize tasks?
### TOUGH LUCK!
This chatbot only has the following to offer!
- Attitude
- Sass
- Frustration
- Wasted time
- ~~Ease of learning~~ *this one's debatable*
- Text-based <sup>~~One day he'll become GUI and you won't see it coming~~</sup>

![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/113225702/ca5f6cf6-ceca-44a9-be2d-66e090a94994)
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/113225702/0dd7c664-d837-456b-aba4-66ab0a69e99b)
![image](https://github.com/nus-cs2103-AY2324S1/ip/assets/113225702/def72c02-c42d-4a75-9443-57091ab7df6b)
*If you ask nicely enough, he might just do as you please.*

### What can he do?
- [x] Manage tasks
- [ ] Remind your sorry butt of deadlines (not yet... but soon....)
### Hear what the developer has to say about this product!!!
> I hate it. It was like the chatbot was fighting me at every commit. Get this thing as far away from me as possible. - Eola-Z
### Learn more about what a sleep-deprived CS2103T student can conjure up in 10 minutes!!! 
```java
public static ParserOutput parseInput(String input) {
    if (input.startsWith("mark")) {
        return ParserOutput.MARK;
    }
    if (input.startsWith("unmark")) {
        return ParserOutput.UNMARK;
    }
    if (input.startsWith("delete")) {
        return ParserOutput.DELETE;
    }
    if (input.startsWith("list")) {
        return ParserOutput.LIST;
    }
    if (input.startsWith("echo")) {
        return ParserOutput.ECHO;
    }
    if (input.startsWith("bye")) {
        return ParserOutput.EXIT;
    }
    if (input.startsWith("find")) {
        return ParserOutput.FIND;
    }
    return ParserOutput.APPEND;
}
```
Now that is some A-list usage of enums right there. ⭐ ⭐ ⭐ 💯 💯 

Other oddities include:
1. Ultra-selective application of OOP
2. Bizarre Javadoc comments
3. (Almost) No JUnit testing!

If you'd like to witness the horror first-hand, [here ya go](https://github.com/Eola-Z/ip)!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Eola-Z committed:
Personalized the chatbot

Eola-Z committed:
Added the Echo functionality to the chatbot

Eola-Z committed:
Added functionality to append and list tasks

Eola-Z committed:
Add a new Task class, and add functionality to toggle tasks between 'done' and 'not done'.

Eola-Z committed:
Create new subclasses of Task to support labels and deadlines; and add functionality to manage the various types of tasks

Eola-Z committed:
Added some test files

Eola-Z committed:
Replace the task array with an arraylist; add functionality to delete tasks

Eola-Z committed:
Add file writing and reading to save tasks added by user

Eola-Z committed:
Merge branch 'branch-Level-7'

Eola-Z committed:
Add parsing for dates and formatted display of dates

Eola-Z committed:
Merge branch 'branch-Level-8'

Eola-Z committed:
Add date parsing at Event/Deadline initialization to catch exceptions

Eola-Z committed:
Add Storage class to read, write and delete task data

Eola-Z committed:
Add Parser to manage text input by the user

Eola-Z committed:
Remove redundant constants

Eola-Z committed:
Merge branch 'Branch-More-OOP'

Eola-Z committed:
Package all classes into one 'duke' package

Eola-Z committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-Gradle

Eola-Z committed:
Set up Gradle to run Duke from it

Eola-Z committed:
Merge branch 'branch-Gradle'

Eola-Z committed:
Add JUnit tests for Parser and ToDo classes

Eola-Z committed:
Merge branch 'branch-JUnit'

Eola-Z committed:
Add JavaDoc comments

Eola-Z committed:
Modify Storage.java to comply with coding standard

Eola-Z committed:
Add search feature to Duke

Eola-Z committed:
Merge branch 'branch-A-JavaDoc'

Eola-Z committed:
Merge branch 'branch-A-CodingStandard'

Eola-Z committed:
Merge branch 'branch-Level-9'

lunaflight commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+     * Common text elements to be reused by the chatbot, such as the chatbot name,
+     * and the horizontal line element.
+     */
+    private static final String name = "JOHNATHAN CENATOR\n";
+    private static final String horizontalLine = "-------------------------------\n";
```
What do you think of all-caps snake case for these constants?

lunaflight commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,164 @@
+    /**
+     * Loads tasks from data file if it exists
+     */
+    public Storage() {
```
This constructor seems quite long with a lot of logic.

What do you think about abstracting the creation of the directory and writing of tasks into a private method in the same class?

lunaflight commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,164 @@
+
+        try {
+            FileReader fr = new FileReader("./data/duke.txt");
+            int c;
+            String savedTasks = "";
+            while ((c=fr.read()) != -1) {
+                savedTasks += (char) c;
+            }
```
This part seems a lot like magic, including a cast of an `int` to a `char` as well as an assignment operator's return value being used.

Do you think that this part could be made more maintainable?

lunaflight commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,164 @@
+            } else if (taskToSave instanceof Deadline) {
+                taskType = "D/";
+                taskAppendices = "/"  + taskToSave.description + "/" + ((Deadline) taskToSave).by;
+            } else if (taskToSave instanceof  Event) {
```
Heads up - there are 2 whitespaces before Event.

lunaflight commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,101 @@
+     * @param input the text typed by the user
+     * @return An enum which indicates what the UI should do next
+     */
+    public static ParserOutput parseInput(String input) {
```
I love this parser!

lunaroddity replied:
Agreed! Very clean :D

lunaflight reviewed (commented):
Looks good! Just a few nits :+1:

Eola-Z committed:
Add checkstyle module in Duke

Eola-Z committed:
Merge branch 'branch-A-Checkstyle'

lunaroddity commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * Create task based on description and deadline
+     *
+     * @param description The description of the task
```
I noticed that your parameter descriptions were missing punctuation behind them throughout your code, what do you think of adding them?

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+                name +
+                "COMING DID YOU!?\n" +
+                horizontalLine);
+    }
```
What do you think of putting the breaks after the operator instead?

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+     */
+    private static void append(Task task) {
+        taskStorage.appendTask(task);
+        System.out.print(horizontalLine + "YOU WANT TO " + task + "?\nSURE, WHATEVER.\n" + horizontalLine);
```
Any reason why you decided to keep this in one line instead of 
```java
System.out.print(horizontalLine +
                "YOU DIDN'T SEE\n" +
                name +
                "COMING DID YOU!?\n" +
                horizontalLine);
```
like the above?

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+                    "event {task} /from {time} /to {time}\n" +
+                    horizontalLine);
+        } catch (DateTimeParseException e) {
+            System.out.print(horizontalLine + "Date format should be yyyy-mm-dd\n"
```
I liked how strong the personality of your bot is! Have you considered making the character of this reply more like the rest?

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+    /**
+     * Attempts to mark a task in the task array
+     *
+     * @param toMark
```
Perhaps add a short description of this parameter?

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+        try {
+            Task task = taskStorage.get(Parser.parseMark(toMark));
+            if (task == null) throw new NullPointerException();
+            if (task.isDone) throw new IllegalArgumentException();
```
What do you think of doing this instead?
```suggestion
            if (task.isDone) {
              throw new IllegalArgumentException();
            }
```

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+
+    /**
+     * Attempts to delete a task from the task array
+     * @param toDelete
```
Perhaps add a short description of this parameter?

lunaroddity commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,277 @@
+        greet();
+        Scanner textInput = new Scanner(System.in);
+        Status botStatus = Status.RUNNING;
+
```
You have an extra newline here

lunaroddity commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @param description
+     * @param from
+     * @param to
```
Noticed this issue in other places in your code as well, perhaps add short descriptions for each of these parameters?

lunaroddity reviewed (commented):
Just some nits here and there but otherwise LGTM!

Eola-Z committed:
Add JavaFx functionality to Duke

Eola-Z committed:
Merge branch 'branch-JavaFX'

Eola-Z committed:
Configure Duke to use the JavaFX GUI

Eola-Z committed:
Merge branch 'branch-Level-10'

Eola-Z committed:
Set main class to Launcher for creating .jar files

Eola-Z committed:
Add assertions to core classes

Assertions allow us to define assumptions about the program state so that the runtime can verify them. Assertion failures indicate possible bugs in the code.

Implementing assertions in the code allows us to affirm these assumptions to detect potential bugs.

Assertions are added to the Storage class to ensure the versatility of the file selection; the directory should be in the right directory and have the correct name, no other files in the directory, etc. We also check that the search functions in the Storage class are thorough in their search.
Similarly, assertions are added to the Duke class to ensure text fields are properly cleared after each query by the user.

Eola-Z committed:
Make improvements to code quality

Code quality is essential in ensuring readable code. Sometimes good quality code may be easier to compile and can run faster.

Let's refactor the Duke class so that
* Operators are more consistently formatted in the code
* If blocks are properly surrounded with braces

Let's also refactor the Storage class to
* Extract long methods into shorter, readable methods
* Introduce whitespace to make code blocks easier to differentiate

Eola-Z committed:
Merge pull request #2 from Eola-Z/branch-A-Assertions

Add assertions to core classes

Eola-Z committed:
Merge branch 'master' into branch-A-CodeQuality

Eola-Z committed:
Merge pull request #1 from Eola-Z/branch-A-CodeQuality

Make improvements to code quality

Eola-Z committed:
Clean up code, remove unused imports, and fix typos

Eola-Z committed:
Add tags as variable to Task class

Tagging is another useful way to search for tasks, and it may be useful to implement in the tP.

Let's eventually:
*Allow the constructor to include tags
*Update Storage to read and write tags
*Enable the parser to detect tagging functions (add, delete, search by tag)

Eola-Z committed:
Merge branch 'branch-C-Tagging'

Eola-Z committed:
Add reading and writing tags in Storage class

Eola-Z committed:
Merge branch 'branch-C-Tagging'

Eola-Z committed:
Enable screen resizing

Eola-Z committed:
Create user guide

Eola-Z committed:
Add tagging functionality in the application

Eola-Z committed:
Fix typos in readme.md

Eola-Z committed:
Fix incorrect naming of Ui.png

Eola-Z committed:
Catch and fix a major bug during testing

Storage was not implemented to correctly handle task text that did not contain and tags. This has now been fixed.

Eola-Z committed:
Add more JUnit tests

Eola-Z committed:
Fix an issue with the .jar file crashing on Linux
